# Data Cleaning and Exploration


## Table of Contents:
1. [Important Information](http://localhost:8888/notebooks/Documents/DataSci/Outsider-Music-Linguistic-Analysis/data_cleaning_and_exploration.ipynb#Important-Information)    
2. [Unpickle the Data](http://localhost:8888/notebooks/Documents/DataSci/Outsider-Music-Linguistic-Analysis/data_cleaning_and_exploration.ipynb#Unpickle-the-Data)    
3. [Initial Data Exploration](http://localhost:8888/notebooks/Documents/DataSci/Outsider-Music-Linguistic-Analysis/data_cleaning_and_exploration.ipynb#Initial-Data-Exploration)    
4. [Data Cleaning](http://localhost:8888/notebooks/Documents/DataSci/Outsider-Music-Linguistic-Analysis/data_cleaning_and_exploration.ipynb#Data-Cleaning)    

## Important Information
### Prior to the Creation of this Script:
I have scraped the list of musicians from Wikipedia using wiki_musicians.py.  I manually cleaned the text file to remove double quotations ("") and parentheticals like "performer," "musician," and "singer," since it was manageable with just 111 artists.

I then retrieved lyrics from these artists from Genius.com using lyricsgenius_requests.ipynb, where 70+ artist-queries successfully returned JSON files.

Lastly, I created and pickled the lyrics data frame (or rough_df.pkl) in the load_json_to_df.ipynb script, where an additional 8 artists could not be successfully loaded, due to entirely null lyrical entries.  This pickled data frame remains in my local repository.

The steps above were broken into so many scripts to avoid re-running time-intensive queries.


### The Data Sharing Problem:
Because the musician files and the pickled data frame are somewhat large and the data is fairly accessible (and fairly static in nature) through the Genius API, I have decided not to publish it but to maintain snippets of the data in presentation, as well as the Crispin Glover JSON included in `data_samples`, for reference to the JSON file structure.

## Unpickle the Data

In [1]:
import pickle
f = open('data/rough_df.pkl', 'rb')
lyricsdf = pickle.load(f)
f.close()

In [2]:
lyricsdf.head()

artist                    title  \
0  Abner Jay   Bring It When You Come   
1  Abner Jay                  Cocaine   
2  Abner Jay            Cocaine Blues   
3  Abner Jay  Don’t Mess with Me Baby   
4  Abner Jay                  I cried   

                                              lyrics  
0  Bring It When You Come Lyrics\nTerrible things...  
1  Cocaine Lyrics\nFolk music is high-class music...  
2  Cocaine Blues LyricsCocaine, Cocaine\nRunning ...  
3  Don’t Mess with Me Baby Lyrics\nTerrible thing...  
4  I cried LyricsI, I cried\nI cried in the ni-ig...

## Initial Data Exploration
Before cleaning, I'd like to see the status of the data: how many artists are there, how many songs, how many lyrics are empty strings, how many are duplicates, etc.

In [3]:
import nltk
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:
lyricsdf.describe()

artist    title lyrics
count            3388     3388   3388
unique             63     3359   3187
top     Wesley Willis  Lullaby       
freq              561        3    202

It looks like there are 3,388 lyric entries, with 63 unique artists. I sense there are a number of "null" lyrics, given that nearly 200 entries are not unique.  This may also be explained by duplicates (though I wouldn't expect exact duplicates for non-null lyrics), given that Wesley Willis's "Lullaby" is somehow the "top" song.

## Data Cleaning